crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -392,6 +392,27 @@ impl<InstalledPackages: InstalledPackagesProvider> ResolverState<InstalledPackag
                         }
                     }
 
+                    // If another fork had the same resolution, merge into that fork instead.
+                    if let Some(existing_resolution) = resolutions
+                        .iter_mut()
+                        .find(|existing_resolution| resolution.same_graph(existing_resolution))
+                    {
+                        let ResolverMarkers::Fork(existing_markers) = &existing_resolution.markers
+                        else {
+                            panic!("A non-forking resolution exists in forking mode")
+                        };
+                        let mut new_markers = existing_markers.clone();
+                        new_markers.or(resolution
+                            .markers
+                            .fork_markers()
+                            .expect("A non-forking resolution exists in forking mode")
+                            .clone());
+                        existing_resolution.markers = normalize(new_markers, None)
+                            .map(ResolverMarkers::Fork)
+                            .unwrap_or(ResolverMarkers::Universal);
+                        continue 'FORK;
+                    }
```

Comment:
> Firstly, is it possible for more than one duplicate resolution to exist at any given point in time? If so, this would I believe only find one of them. But, I do not think this is case, since this is run for every resolution before it is "saved." So it should never be the case that more than one duplicate resolution appears.

We fork every time we see conflicting markers, but in many of those cases the requirements themselves are not conflicting (say `numpy >= 1.16` for one and `numpy >= 1.19` for the other). When forking, we can't yet know whether we'll find a compatible `numpy` for both of. I've also seen cases where we end up rejecting the package version we forked on in both branches, removing the conflicting requirements. By copying over preferences from previous forks, we try to coerce two forks to resolving the same package version. Basically, our strategy is to fork often to avoid failing on avoidable conflicts, but still having a solution with as few divergences as possibles.

Re perf: I agree that this is potentially costly, but i think we have to do this to get desirable resolution. We have some short-cuts that we get from std that makes this cheaper: When two forks have a different number of packages, the check is a single `usize` comparison. We also usually have a small number of forks (and the more specific a fork is, the more likely it is we skip future fork points because we're already more specific), so i see this more as a fixed cost of maybe 10^2/2=50 checks. There are of course pathological cases, for those i think we just have to be a bit slow here avoid redundant forks in the lockfile.